_posts/2015/2015-06-29-git-migrate.markdown

```diff
@@ -0,0 +1,153 @@
+
+`git reset origin/master --hard`
+
+![Reset master back to its original position](https://cloud.githubusercontent.com/assets/19977/8390501/bbf39b24-1c4c-11e5-8d7a-b1cf7b4bcb5b.png)
+
+Notice that `HEAD` and `master` are back to their original position, but `new-branch` continues to point to the latest commit.
+
+Now I can simply check out the `new-branch` and continue my work.
+
+![Checkout new-branch and we are done](https://cloud.githubusercontent.com/assets/19977/8390502/bbf69180-1c4c-11e5-920a-7991e341eefa.png)
+
+This moves `HEAD` to the new branch and now I can continue to work on the branch as if I had never made this mistake. Yay!
+
+Here's the set of commands that I ran all together.
+
+```bash
+git branch new-branch
+git reset origin/master --hard
+git checkout new-branch
+```
+
+### Fixing up a non-master branch
+
+![The wrong branch](https://cloud.githubusercontent.com/assets/19977/8369613/48019364-1b71-11e5-9a28-b748a2802ed7.png)
+
+This case is a bit more complicated. Here I have a branch named `wrong-branch` that is my current branch. But I thought I was working in the `master` branch. I make two commits in this branch by `mistake` which causes this fine mess.
+
+![A fine mess](https://cloud.githubusercontent.com/assets/19977/8369614/4801b83a-1b71-11e5-898a-4c116e83b749.png)
+
+What I want here is to migrate commits `E` and `F` to a new branch off of `master`. Here's the set of commands.
+
+Let's walk through these steps one by one. Not to worry, as before, I create a new branch.
+
+`git branch new-branch`
+
+![Always a new branch](https://cloud.githubusercontent.com/assets/19977/8390535/e6a2c114-1c4d-11e5-8fd7-0911a8238334.png)
+
+Again, just like before, I reset the current branch to the state of the current branch as it exists on the server.
+
+`git reset origin/wrong-branch --hard`
+
+![Reset the current branch](https://cloud.githubusercontent.com/assets/19977/8369612/47fe6374-1b71-11e5-92dd-26ddf71d5a7a.png)
+
+But now, I need to move the `new-branch` onto `master`.  
+
+`git rebase --onto master origin/wrong-branch new-branch`
+
+![Final result](https://cloud.githubusercontent.com/assets/19977/8382092/06f71640-1be5-11e5-9f90-2b433bd6ffd8.png)
+
+The `git rebase` command is a great way to move (well, actually you replay commits, but that's a story for another day) commits onto other branches. The handy `--onto` flag makes it possible to specify a range of commits to move elsewhere. [Pivotal Labs has a helpful post](http://pivotallabs.com/git-rebase-onto/) that describes this option in more detail.
+
+So in this case, I moved commits `E` and `F` because they are the ones between `origin/wrong-branch` exclusive and `new-branch` inclusive
+
+Here's the set of command I ran all together.
+
+```bash
+git branch new-branch
+git reset origin/wrong-branch --hard
+git rebase --onto master origin/wrong-branch new-branch
+```
```

Comment:
>  and the two-argument form of it even more unintuitive. So it's nice that the latter is not necessary here.

But it looks like you used the 2-arg form. Did you mean the 3-arg form is not necessary here?